core/src/main/java/com/google/zxing/qrcode/decoder/DataMask.java

```diff
@@ -29,135 +29,126 @@
    */
-  private static final class DataMask000 extends DataMask {
+  DATA_MASK_000() {
     @Override
     boolean isMasked(int i, int j) {
       return ((i + j) & 0x01) == 0;
     }
-  }
+  },
 
   /**
    * 001: mask bits for which x mod 2 == 0
    */
-  private static final class DataMask001 extends DataMask {
+  DATA_MASK_001() {
     @Override
     boolean isMasked(int i, int j) {
       return (i & 0x01) == 0;
     }
-  }
+  },
 
   /**
    * 010: mask bits for which y mod 3 == 0
    */
-  private static final class DataMask010 extends DataMask {
+  DATA_MASK_010() {
     @Override
     boolean isMasked(int i, int j) {
       return j % 3 == 0;
     }
-  }
+  },
 
   /**
    * 011: mask bits for which (x + y) mod 3 == 0
    */
-  private static final class DataMask011 extends DataMask {
+  DATA_MASK_011() {
     @Override
     boolean isMasked(int i, int j) {
       return (i + j) % 3 == 0;
     }
-  }
+  },
 
   /**
    * 100: mask bits for which (x/2 + y/3) mod 2 == 0
    */
-  private static final class DataMask100 extends DataMask {
+  DATA_MASK_100() {
     @Override
     boolean isMasked(int i, int j) {
       return (((i / 2) + (j /3)) & 0x01) == 0;
     }
-  }
+  },
 
   /**
    * 101: mask bits for which xy mod 2 + xy mod 3 == 0
    */
-  private static final class DataMask101 extends DataMask {
+  DATA_MASK_101() {
     @Override
     boolean isMasked(int i, int j) {
       int temp = i * j;
       return (temp & 0x01) + (temp % 3) == 0;
     }
-  }
+  },
 
   /**
    * 110: mask bits for which (xy mod 2 + xy mod 3) mod 2 == 0
    */
-  private static final class DataMask110 extends DataMask {
+  DATA_MASK_110() {
     @Override
     boolean isMasked(int i, int j) {
       int temp = i * j;
       return (((temp & 0x01) + (temp % 3)) & 0x01) == 0;
     }
-  }
+  },
 
   /**
    * 111: mask bits for which ((x+y)mod 2 + xy mod 3) mod 2 == 0
    */
-  private static final class DataMask111 extends DataMask {
+  DATA_MASK_111() {
     @Override
     boolean isMasked(int i, int j) {
       return ((((i + j) & 0x01) + ((i * j) % 3)) & 0x01) == 0;
     }
+  };
+
+  // End of enum constants.
+
+
+  /**
+   * <p>Implementations of this method reverse the data masking process applied to a QR Code and
+   * make its bits ready to read.</p>
+   *
+   * @param bits representation of QR Code bits
+   * @param dimension dimension of QR Code, represented by bits, being unmasked
+   */
+  final void unmaskBitMatrix(BitMatrix bits, int dimension) {
+    for (int i = 0; i < dimension; i++) {
+      for (int j = 0; j < dimension; j++) {
+        if (isMasked(i, j)) {
+          bits.flip(j, i);
+        }
+      }
+    }
+  }
+
+  abstract boolean isMasked(int i, int j);
+
+  /**
+   * @param reference a value between 0 and 7 indicating one of the eight possible
+   * data mask patterns a QR Code may use
+   * @return DataMask encapsulating the data mask pattern
+   */
+  static DataMask forReference(int reference) {
```

Comment:
I was going to say we can remove this method because enum exposes values(). But it does very technically change the exception for bad values. Still it is an internal API and really this error condition ought to be caught by the framework. I'm ok removing it and updating all callers to use values()